test/core/iomgr/tcp_client_posix_test.cc

```diff
@@ -207,27 +211,84 @@ void test_fails(void) {
 
 void test_connect_cancellation_succeeds(void) {
   gpr_log(GPR_ERROR, "---- starting test_connect_cancellation_succeeds() ----");
+  auto target_ipv6_addr_uri = *grpc_core::URI::Parse(absl::StrCat(
+      "ipv6:[::1]:", std::to_string(grpc_pick_unused_port_or_die())));
+  auto target_ipv4_addr_uri = *grpc_core::URI::Parse(absl::StrCat(
+      "ipv4:127.0.0.1:", std::to_string(grpc_pick_unused_port_or_die())));
   grpc_resolved_address resolved_addr;
-  struct sockaddr_in* addr =
-      reinterpret_cast<struct sockaddr_in*>(resolved_addr.addr);
   int svr_fd;
   grpc_closure done;
   grpc_core::ExecCtx exec_ctx;
-
-  memset(&resolved_addr, 0, sizeof(resolved_addr));
-  resolved_addr.len = static_cast<socklen_t>(sizeof(struct sockaddr_in));
-  addr->sin_family = AF_INET;
-
+  bool tried_ipv4 = false;
+  ASSERT_TRUE(grpc_parse_uri(target_ipv6_addr_uri, &resolved_addr));
+  auto try_bind = [&](int sock) {
+    if (sock < 0 || bind(sock, reinterpret_cast<sockaddr*>(resolved_addr.addr),
+                         resolved_addr.len) != 0) {
+      return false;
+    }
+    return true;
```

Comment:
Could be a one-liner `return (sock >= 0 && bind(...) == 0);`

Reply:
Fixed.